File: Graphs/Detect-Cycle-in-Graph.js
"use strict";
const Graph = require('./Graph.js');

/**
 * For each node, we start a recursive call with detectCycleRec.
 * The function maintains two stacks (not DS one) -> recNodes and visited
 * 
 * recNodes -> The vertices that have been traversed in the current recursion are added to it  and
 * visited -> keeps a record of all the nodes that have been traversed regardless of the recursive call.
 * 
 * For a cycle to occur, we must reach a node that was already present in the recursion stack. 
 * If the recursion ends and no such node is found, the stack is reset again and the next iteration 
 * of detectCycle starts.
 */
/**
 * Runtime Complexity: O(V + E)
 * Space Complexity: O(V) ??
 */

function detectCycleRec(graph, i,visited, recNodes) {
    // Check if current node is being visited in the same recursive call
    if(visited[i] === false) {
        // Set recursive array and visited to true
        visited[i] = true;
        recNodes[i] = true;

        var adjacent;
        let adjacentNode = graph.list[i].getHead();
        while (adjacentNode !== null ) {
            // Access adjacent node and repeat algorithm
            adjacent = adjacentNode.data;
            if ((!visited[adjacent]) && detectCycleRec(graph, adjacent, visited, recNodes)) {
                return true;  // Loop found
            } else if(recNodes[adjacent]) {
                return true;
            }
            adjacentNode = adjacentNode.nextElement;
        }
    }
    recNodes[i] = false;
    return false;
}

function detectCycle(graph) {
    var num_of_vertices = graph.vertices;
    //Boolean Array to hold the history of visited nodes (by default-false)
    //Make a node visited whenever you traverse it
     var visited = [];

    //Boolean array of vertices which will recursively called
     var recNodes = [];

    for(var i = 0; i < num_of_vertices; i++)
    {
        visited[i] = false;
        recNodes[i] = false;
    }

    for (var i = 0; i < num_of_vertices; i++) { // Recursive function called
        if (detectCycleRec(graph, i, visited, recNodes)) {
            return true;
            // If cycle detected, return true
        }
    }
    return false;
}


let graph = new Graph(4);
graph.addEdge(0, 1);
graph.addEdge(0, 3);
graph.addEdge(1, 2);
graph.addEdge(2, 0);

graph.printGraph();

console.log(detectCycle(graph)); // true

File: Graphs/DetectCycleInGraph.java
package Graphs;

public class DetectCycleInGraph {
    public static boolean detectCycle(Graph g) {
        int numOfVertices = g.vertices;

        // To hold the history of visited nodes
        boolean[] visited = new boolean[numOfVertices];
        // holds a flag if node is currently in Stack or not
        boolean[] stackFlag = new boolean[numOfVertices];

        for (int i = 0; i < numOfVertices; i++) {
            if (cyclic(g, i, visited, stackFlag)) {
                return true;
            }
        }
        return false;
    }

    public static boolean cyclic(Graph g, int v, boolean[] visited, boolean[] stackFlag) {
        // If node is already in stack, cycle is found
        if(stackFlag[v]) {
            return true;
        }

        //if it is already visited (and not in Stack) then there is no cycle
        if (visited[v]) {
            return false;
        }

        visited[v] = true;
        stackFlag[v] = true;

        // check adjacency list of the node
        DoublyLinkedList<Integer>.Node temp = null;
        if (g.adjacencyList[v] != null)
            temp = g.adjacencyList[v].headNode;

        while (temp != null) {
            //run cyclic function recursively on each outgowing path
            if(cyclic(g, temp.data, visited, stackFlag)){
                return true;
            }
            temp = temp.nextNode;
        }
        stackFlag[v] = false;

        return false;
    }

    public static void main(String args[]) {
        Graph g1 = new Graph(4);
        g1.addEdge(0,1);
        g1.addEdge(1,2);
        g1.addEdge(1,3);
        g1.addEdge(3,0);
        g1.printGraph();
        System.out.println(detectCycle(g1));

        System.out.println();
        Graph g2 = new Graph(3);
        g2.addEdge(0,1);
        g2.addEdge(1,2);
        g2.printGraph();
        System.out.println(detectCycle(g2));
  }
}
